### PR TITLE
Derive `Debug` for `SessionStateBuilder`, adding `Debug` requirements to fields

### PR DESCRIPTION
**Pull request by AnthonyZhOon**
## Which issue does this PR close?
Progress on #12555 

## Rationale for this change

To make configuration easier to use by providing debug output.

## What changes are included in this PR?

1. Require `Debug` for `CatalogProvider`, `CatalogProviderList`, `UrlTableFactory`.
2. Require `Debug` for `ExprPlanner`, `QueryPlanner`.
3. Require `Debug` for `TableFunctionImpl`
4. Require `Debug` for `SerializerRegistry`
5. Require `Debug` for `FunctionFactory`
6. `#derive` debug on necessary classes
7. Implement `Debug` on `SessionStateBuilder` by deriving.

## Are these changes tested?

By the compiler and CI.

## Are there any user-facing changes?

The additional `Debug` requirement is a user-facing API change on the traits affected.

### CODE REVIEW

**Comment by AnthonyZhOon** [comment]
Needs the `api-change` tag. I'm also fine with splitting this PR into smaller ones

**Comment by AnthonyZhOon** [comment]
Now the impl Debug for `SessionStateBuilder` resembles that of `SessionState` but includes all fields. Also grouped fields by relevance

File: datafusion/core/src/catalog_common/information_schema.rs [hunk 1]
```diff
@@ -70,20 +71,11 @@ impl InformationSchemaProvider {
...
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 struct InformationSchemaConfig {
     catalog_list: Arc<dyn CatalogProviderList>,
 }
 
-impl Debug for InformationSchemaConfig {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("InformationSchemaConfig")
-            // TODO it would be great to print the catalog list here
```
Comment (alamb) [inline datafusion/core/src/catalog_common/information_schema.rs:81]: 🎉

File: datafusion/core/src/execution/session_state.rs [hunk 2]
```diff
@@ -177,23 +177,23 @@ impl Debug for SessionState {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("SessionState")
             .field("session_id", &self.session_id)
-            .field("analyzer", &"...")
+            .field("config", &self.config)
```
Comment (alamb) [inline datafusion/core/src/execution/session_state.rs:180]: I wonder if we could simply `#derive(Debug)` for SessionState now? 

And perhaps we could provide a `impl Display for SessionState` that had a more readable format (like that shows session_id and deviation from default configuration, for example 🤔 )

**Review by alamb — APPROVED** [review]
Thank you @AnthonyZhOon  -- this looks like a great improvement to me. I agree that the actual utility of the `Debug` impl is debatable, but at least now it is possible

**Comment by alamb** [comment]
Thanks again @AnthonyZhOon

**Pull request merged**